coalib/core/Core.py

```diff
@@ -235,11 +237,18 @@ def _schedule_bears(self, bears):
                 logging.debug('Scheduled {!r} (tasks: {})'.format(bear,
                                                                   len(tasks)))
 
+                # Cleanup bears without tasks after all bears had the chance to
+                # schedule their tasks. Not doing so might stop the run too
+                # early, as the cleanup is also responsible for stopping the
+                # event-loop when no more tasks do exist.
                 if not tasks:
-                    # We need to recheck our runtime if something is left to
-                    # process, as when no tasks were offloaded the event-loop
-                    # could hang up otherwise.
-                    self._cleanup_bear(bear)
+                    bears_without_tasks_to_cleanup.append(bear)
```

Comment:
Can we keep this condition before looping over the tasks ?

Something like:
```
if not tasks:
    bears_without_tasks_to_cleanup.append(bear)
    continue
```

Might not save too many clock cycles, but it will make code cleaner.

Reply:
Yeah why not, if you like I would add then an additional debug-message `BearXYZ scheduled no tasks.`, so we still keep the information that a bear has been processed :+1:

Reply:
Hm just thinking if this should be done in a separate commit... This commit diff is smaller when not shifting the logic away^^

Reply:
IMO this can be done in the same commit as we are adding this condition in this commit. It doesn't make sense to add something in first commit and change the same in subsequent one especially in same PR.

Reply:
And there won't be a case where we only want to rollback one of the commit and not the other one. So, its better to keep single commit.

Reply:
> as we are adding this condition in this commit

which condition? The `if` is already there (`if not tasks`)

Reply:
Oh my bad. Can be done in different commit.